File: client-extensions/ruby/agent_auth.rb
# frozen_string_literal: true
require 'json'
require 'uri'
require 'net/http'

module AgentTrunk
  class AgentAuthError < StandardError; end
  # No credential persistence or automatic retries. Humans approve externally.
  class AgentRegistration
    attr_reader :resource, :issuer, :guide
    def self.origin(value)
      u = URI.parse(value)
      raise AgentAuthError, 'invalid_origin' unless u.is_a?(URI::HTTPS) && u.host && !u.userinfo && ['', '/'].include?(u.path) && !u.query && !u.fragment
      u.path = ''; u.to_s
    rescue URI::InvalidURIError, TypeError
      raise AgentAuthError, 'invalid_origin'
    end
    def self.endpoint(value, issuer)
      u = URI.parse(value)
      base = u.dup; base.path = ''; base.query = nil; base.fragment = nil
      raise AgentAuthError, 'untrusted_endpoint' unless base.to_s == issuer && !u.userinfo && !u.fragment
      value
    rescue URI::InvalidURIError, TypeError
      raise AgentAuthError, 'invalid_discovery'
    end
    def self.secret(value)
      raise AgentAuthError, 'invalid_credentials' unless value.is_a?(String) && !value.empty? && value.bytesize <= 32768 && !value.match?(/\s/)
      value
    end
    def self.request(url, body=nil, form: false)
      u = URI.parse(url)
      req = body.nil? ? Net::HTTP::Get.new(u) : Net::HTTP::Post.new(u)
      if body
        req['Content-Type'] = form ? 'application/x-www-form-urlencoded' : 'application/json'
        req.body = form ? URI.encode_www_form(body) : JSON.generate(body)
      end
      bytes = +''
      http = Net::HTTP.new(u.host, u.port, nil)
      http.use_ssl = true; http.open_timeout = 15; http.read_timeout = 15; http.write_timeout = 15; http.max_retries = 0
      http.start do |connection|
        connection.request(req) do |response|
          raise AgentAuthError, "request_rejected:#{response.code}" unless response.code.to_i.between?(200,299)
          response.read_body do |chunk|
            bytes << chunk
            raise AgentAuthError, 'response_too_large' if bytes.bytesize > 131072
          end
        end
      end
      result = JSON.parse(bytes)
      raise AgentAuthError, 'invalid_response' unless result.is_a?(Hash)
      result
    rescue AgentAuthError
      raise
    rescue StandardError
      raise AgentAuthError, 'network_or_response_failure', cause: nil
    end
    def self.discover(resource='https://api.agenttrunk.ai')
      api = origin(resource); meta = request(api+'/.well-known/oauth-protected-resource')
      servers = meta['authorization_servers']
      raise AgentAuthError, 'invalid_discovery' unless meta['resource'] == api && servers.is_a?(Array) && servers.length == 1
      issuer = origin(servers[0]); server = request(issuer+'/.well-known/oauth-authorization-server')
      raise AgentAuthError, 'registration_unavailable' unless server['issuer'] == issuer && server.dig('agent_auth','identity_types_supported')&.include?('service_auth')
      new(api,issuer,server)
    end
    def initialize(resource,issuer,server)
      @resource, @issuer = resource, issuer
      auth = server.fetch('agent_auth')
      @guide = self.class.endpoint(auth['skill'],issuer)
      @identity_endpoint = self.class.endpoint(auth['identity_endpoint'],issuer)
      @claim_endpoint = self.class.endpoint(auth['claim_endpoint'],issuer)
      @token_endpoint = self.class.endpoint(server['token_endpoint'],issuer)
    end
    def start(email)
      raise AgentAuthError, 'invalid_email' unless email.is_a?(String) && email.bytesize <= 254 && email.match?(/\A[^\s@]+@[^\s@]+\.[^\s@]+\z/)
      result = self.class.request(@identity_endpoint,{'type'=>'service_auth','login_hint'=>email})
      claim = result.fetch('claim'); token = self.class.secret(claim['token'])
      attempt = claim['attempt'] || self.class.request(@claim_endpoint,{'type'=>'service_auth','login_hint'=>email,'claim_token'=>token})['attempt']
      {'claim_token'=>token,'verification_uri'=>self.class.endpoint(attempt['verification_uri'],@issuer)}
    end
    def complete(claim_token,user_code)
      raise AgentAuthError, 'invalid_user_code' unless user_code.is_a?(String) && user_code.match?(/\A[A-Za-z0-9-]{4,32}\z/)
      identity(self.class.request(@claim_endpoint+'/complete',{'claim_token'=>self.class.secret(claim_token),'user_code'=>user_code})['identity'])
    end
    def exchange(value)
      result = self.class.request(@token_endpoint,{'grant_type'=>'urn:ietf:params:oauth:grant-type:jwt-bearer','assertion'=>self.class.secret(value['assertion']),'resource'=>@resource},form:true)
      ttl = result['expires_in']
      raise AgentAuthError, 'invalid_credentials' unless result['token_type'].to_s.downcase == 'bearer' && ttl.is_a?(Numeric) && ttl.finite? && ttl > 0
      {'access_token'=>self.class.secret(result['access_token']),'expires_in'=>ttl}
    end
    def refresh(value)
      identity(self.class.request(@identity_endpoint,{'type'=>'refresh','refresh_token'=>self.class.secret(value.dig('refresh_token','value'))})['identity'])
    end
    private
    def identity(value)
      result = {'assertion'=>self.class.secret(value&.fetch('assertion',nil))}
      result['refresh_token'] = {'value'=>self.class.secret(value.dig('refresh_token','value'))} if value['refresh_token']
      result
    end
  end
end

File: client-extensions/go/agent_auth.go
// Package agentauth implements human-approved registration; callers own secret storage.
package agentauth

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

type Identity struct {
	Assertion    string        `json:"assertion"`
	RefreshToken *RefreshToken `json:"refresh_token,omitempty"`
}
type RefreshToken struct {
	Value string `json:"value"`
}
type Attempt struct {
	ClaimToken      string
	VerificationURI string
}
type Token struct {
	AccessToken string
	ExpiresIn   float64
}
type Registration struct {
	Resource, Issuer, Guide string
	identity, claim, token  string
	client                  *http.Client
}

func origin(value string) (string, error) {
	u, e := url.Parse(value)
	if e != nil || u.Scheme != "https" || u.Host == "" || u.User != nil || (u.Path != "" && u.Path != "/") || u.RawQuery != "" || u.Fragment != "" {
		return "", errors.New("invalid origin")
	}
	return "https://" + u.Host, nil
}
func endpoint(value, issuer string) (string, error) {
	u, e := url.Parse(value)
	if e != nil || u.Scheme+"://"+u.Host != issuer || u.User != nil || u.Fragment != "" {
		return "", errors.New("untrusted endpoint")
	}
	return value, nil
}
func secret(value string) error {
	if value == "" || len(value) > 32768 || strings.ContainsAny(value, " \t\r\n") {
		return errors.New("invalid credentials")
	}
	return nil
}
func (r *Registration) request(ctx context.Context, address string, body any, form bool) (map[string]json.RawMessage, error) {
	var data []byte
	var err error
	method := "GET"
	contentType := "application/json"
	if body != nil {
		method = "POST"
		if form {
			data = []byte(body.(url.Values).Encode())
			contentType = "application/x-www-form-urlencoded"
		} else {
			data, err = json.Marshal(body)
			if err != nil {
				return nil, errors.New("invalid request")
			}
		}
	}
	req, err := http.NewRequestWithContext(ctx, method, address, bytes.NewReader(data))
	if err != nil {
		return nil, errors.New("invalid request")
	}
	if body != nil {
		req.Header.Set("Content-Type", contentType)
	}
	response, err := r.client.Do(req)
	if err != nil {
		return nil, errors.New("authentication network failure")
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return nil, fmt.Errorf("authentication rejected (HTTP %d)", response.StatusCode)
	}
	data, err = io.ReadAll(io.LimitReader(response.Body, 131073))
	if err != nil || len(data) > 131072 {
		return nil, errors.New("invalid authentication response")
	}
	var result map[string]json.RawMessage
	if json.Unmarshal(data, &result) != nil || result == nil {
		return nil, errors.New("invalid authentication response")
	}
	return result, nil
}
func field[T any](data map[string]json.RawMessage, key string) (T, error) {
	var value T
	raw, ok := data[key]
	if !ok || json.Unmarshal(raw, &value) != nil {
		return value, errors.New("invalid authentication response")
	}
	return value, nil
}
func Discover(ctx context.Context, resource string, client *http.Client) (*Registration, error) {
	api, err := origin(resource)
	if err != nil {
		return nil, err
	}
	c := http.Client{Timeout: 15 * time.Second}
	if client != nil {
		c = *client
	}
	c.CheckRedirect = func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }
	c.Jar = nil
	if c.Timeout == 0 {
		c.Timeout = 15 * time.Second
	}
	r := &Registration{Resource: api, client: &c}
	meta, err := r.request(ctx, api+"/.well-known/oauth-protected-resource", nil, false)
	if err != nil {
		return nil, err
	}
	named, err := field[string](meta, "resource")
	if err != nil || named != api {
		return nil, errors.New("invalid resource")
	}
	servers, err := field[[]string](meta, "authorization_servers")
	if err != nil || len(servers) != 1 {
		return nil, errors.New("invalid discovery")
	}
	r.Issuer, err = origin(servers[0])
	if err != nil {
		return nil, err
	}
	server, err := r.request(ctx, r.Issuer+"/.well-known/oauth-authorization-server", nil, false)
	if err != nil {
		return nil, err
	}
	named, err = field[string](server, "issuer")
	if err != nil || named != r.Issuer {
		return nil, errors.New("invalid issuer")
	}
	auth, err := field[map[string]json.RawMessage](server, "agent_auth")
	if err != nil {
		return nil, err
	}
	types, err := field[[]string](auth, "identity_types_supported")
	if err != nil {
		return nil, err
	}
	found := false
	for _, t := range types {
		if t == "service_auth" {
			found = true
		}
	}
	if !found {
		return nil, errors.New("registration unavailable")
	}
	for key, target := range map[string]*string{"skill": &r.Guide, "identity_endpoint": &r.identity, "claim_endpoint": &r.claim} {
		value, e := field[string](auth, key)
		if e != nil {
			return nil, e
		}
		*target, e = endpoint(value, r.Issuer)
		if e != nil {
			return nil, e
		}
	}
	value, err := field[string](server, "token_endpoint")
	if err != nil {
		return nil, err
	}
	r.token, err = endpoint(value, r.Issuer)
	return r, err
}
func (r *Registration) Start(ctx context.Context, email string) (Attempt, error) {
	if len(email) > 254 || !regexp.MustCompile(`^[^\s@]+@[^\s@]+\.[^\s@]+$`).MatchString(email) {
		return Attempt{}, errors.New("invalid email")
	}
	result, err := r.request(ctx, r.identity, map[string]string{"type": "service_auth", "login_hint": email}, false)
	if err != nil {
		return Attempt{}, err
	}
	claim, err := field[map[string]json.RawMessage](result, "claim")
	if err != nil {
		return Attempt{}, err
	}
	token, err := field[string](claim, "token")
	if err != nil {
		return Attempt{}, err
	}
	if err = secret(token); err != nil {
		return Attempt{}, err
	}
	var attempt map[string]json.RawMessage
	if raw, ok := claim["attempt"]; ok {
		if json.Unmarshal(raw, &attempt) != nil || attempt == nil {
			return Attempt{}, errors.New("invalid attempt")
		}
	} else {
		result, err = r.request(ctx, r.claim, map[string]string{"type": "service_auth", "login_hint": email, "claim_token": token}, false)
		if err != nil {
			return Attempt{}, err
		}
		attempt, err = field[map[string]json.RawMessage](result, "attempt")
		if err != nil {
			return Attempt{}, err
		}
	}
	uri, err := field[string](attempt, "verification_uri")
	if err != nil {
		return Attempt{}, err
	}
	uri, err = endpoint(uri, r.Issuer)
	return Attempt{token, uri}, err
}
func identity(result map[string]json.RawMessage) (Identity, error) {
	v, e := field[Identity](result, "identity")
	if e != nil {
		return v, e
	}
	if e = secret(v.Assertion); e != nil {
		return Identity{}, e
	}
	if v.RefreshToken != nil {
		e = secret(v.RefreshToken.Value)
	}
	return v, e
}
func (r *Registration) Complete(ctx context.Context, claim, code string) (Identity, error) {
	if secret(claim) != nil || !regexp.MustCompile(`^[A-Za-z0-9-]{4,32}$`).MatchString(code) {
		return Identity{}, errors.New("invalid claim or code")
	}
	result, err := r.request(ctx, r.claim+"/complete", map[string]string{"claim_token": claim, "user_code": code}, false)
	if err != nil {
		return Identity{}, err
	}
	return identity(result)
}
func (r *Registration) Exchange(ctx context.Context, id Identity) (Token, error) {
	if err := secret(id.Assertion); err != nil {
		return Token{}, err
	}
	result, err := r.request(ctx, r.token, url.Values{"grant_type": {"urn:ietf:params:oauth:grant-type:jwt-bearer"}, "assertion": {id.Assertion}, "resource": {r.Resource}}, true)
	if err != nil {
		return Token{}, err
	}
	kind, e := field[string](result, "token_type")
	if e != nil || strings.ToLower(kind) != "bearer" {
		return Token{}, errors.New("invalid token type")
	}
	ttl, e := field[float64](result, "expires_in")
	if e != nil || ttl <= 0 || math.IsInf(ttl, 0) || math.IsNaN(ttl) {
		return Token{}, errors.New("invalid token lifetime")
	}
	access, e := field[string](result, "access_token")
	if e != nil {
		return Token{}, e
	}
	return Token{access, ttl}, secret(access)
}
func (r *Registration) Refresh(ctx context.Context, id Identity) (Identity, error) {
	if id.RefreshToken == nil || secret(id.RefreshToken.Value) != nil {
		return Identity{}, errors.New("missing refresh token")
	}
	result, err := r.request(ctx, r.identity, map[string]string{"type": "refresh", "refresh_token": id.RefreshToken.Value}, false)
	if err != nil {
		return Identity{}, err
	}
	return identity(result)
}
